test/testobject-specs.js

```diff
@@ -0,0 +1,213 @@
+    let execStub, fsStatStub;
+    beforeEach(function () {
+      process.env.TESTOBJECT_USERNAME = 'foobar';
+      process.env.TESTOBJECT_API_KEY = 1234;
+      execStub = sinon.stub(teenProcess, 'exec', () => 1);
+      fsStatStub = sinon.stub(fs, 'stat', () => ({
+        mtime: +(new Date()) - 2 * 24 * 60 * 60 * 1000, // Pretend app was last modified 2 days ago
+      }));
+    });
+    afterEach(function () {
+      execStub.restore();
+      fsStatStub.restore();
+    });
+    it('should be rejected if TESTOBJECT_USERNAME is not defined', async function () {
+      process.env.TESTOBJECT_USERNAME = null;
+      await uploadTestObjectApp().should.eventually.be.rejectedWith(/TESTOBJECT_USERNAME/);
+    });
+    it('should be rejected if TESTOBJECT_API_KEY not set', async function () {
+      process.env.TESTOBJECT_API_KEY = null;
+      await uploadTestObjectApp().should.eventually.be.rejectedWith(/TESTOBJECT_API_KEY/);
+    });
+    it('should be rejected if there is a network error', async function () {
+      process.env.TESTOBJECT_USERNAME = 'foobar';
+      await uploadTestObjectApp().should.eventually.equal(1);
+    });
+    it('should call cURL with -u and --data-binary args', async function () {
+      execStub.restore();
+      execStub = sinon.stub(teenProcess, 'exec', (command, args) => {
+        command.should.equal('curl');
+        args[1].should.equal('"foobar:1234"');
+        args[args.length - 1].should.equal('@fakeapp.app');
+      });
+      await uploadTestObjectApp('fakeapp.app');
+    });
+    it('should re-use appId if app was already uploaded earlier', async function () {
+      TestObject._appIdCache['fakeapp.app'] = {
+        id: 2,
+        uploaded: +(new Date()) - 24 * 60 * 60 * 1000,
+      };
+      await uploadTestObjectApp('fakeapp.app').should.eventually.equal(2);
+      delete TestObject._appIdCache['fakeapp.app'];
+    });
+    it('should save uploaded app to cache', async function () {
+      await uploadTestObjectApp('fakeapp.app');
+      const cache = TestObject._appIdCache['fakeapp.app'];
+
+      // Test that the cache recorded it being uploaded within the last 10 seconds 
+      cache.uploaded.should.be.below(+(new Date()) + 1);
+      cache.uploaded.should.be.above(+(new Date()) - 10000);
+      cache.id.should.equal(1);
+    });
+    it('should re-upload app if app was modified after it was uploaded', async function () {
+      TestObject._appIdCache['fakeapp.app'] = {
+        id: 2,
+        uploaded: +(new Date()) - 2 * 24 * 60 * 60 * 1000 - 1, // 2 days ago minus a millisecond
+      };
+      await uploadTestObjectApp('fakeapp.app').should.eventually.equal(1);
+      TestObject._appIdCache['fakeapp.app'].id.should.equal(1);
+    });
+  });
+
+  describe('#overrideWD', function () {
+    let initSpy, promiseChainRemoteSpy, MockWD;
+
+    beforeEach(function () {
+      initSpy = sinon.spy();
+      promiseChainRemoteSpy = sinon.spy();
+
+      MockWD = function MockWD () {
+
```

Comment:
Yeah it's just a fake WD class that doesn't do anything and is used for testing.